src/FlashContainer.php

```diff
@@ -0,0 +1,130 @@
+ *
+ * The {@see init()} method copies data from the internal container into memory,
+ * then clears it. The data is still accessible from memory,
+ * but no longer from internal container.
+ *
+ * This is useful for flash data, i.e. data that should only be accessible
+ * once per request. If a session-specific persistent container is used
+ * as storage, this will become session-based flash data.
+ */
+class FlashContainer implements
+    MutableContainerInterface,
+    ClearableContainerInterface
+{
+    /** @var MutableContainerInterface */
+    protected $data;
+    /** @var string */
+    protected $dataKey;
+    /** @var array */
+    protected $flashData = [];
+
+    /**
+     * @param MutableContainerInterface $data The storage.
+     * @param string $dataKey The key to be used to store data in the storage.
+     */
+    public function __construct(MutableContainerInterface $data, string $dataKey)
+    {
+        $this->data = $data;
+        $this->dataKey = $dataKey;
+    }
+
+    /**
+     * Prepare storage before use.
+     *
+     * Should be called once before accessing data through this class.
+     * Will clear the data for the configured key from the storage.
+     */
+    public function init(): void
+    {
+        $this->flashData = $this->data->has($this->dataKey)
+            ? $this->data->get($this->dataKey)
+            : [];
+
+        $this->purgePersistentData();
+    }
+
+    /**
+     * Clear data from internal storage.
+     */
+    protected function purgePersistentData(): void
+    {
+        $this->data->set($this->dataKey, []);
+    }
+
+    /**
+     * @inheritDoc
+     */
+    public function has($key)
+    {
+        return array_key_exists($key, $this->flashData);
+    }
+
+    /**
+     * @inheritDoc
+     *
+     * Retrieves the value for the specified key from memory.
+     */
+    public function get($key)
+    {
+        if (!array_key_exists($key, $this->flashData)) {
+            throw new NotFoundException(sprintf('Flash data not found for key "%1$s"', $key));
+        }
+
+        return $this->flashData[$key];
+    }
+
+    /**
+     * @inheritDoc
+     *
+     * Assigns the given value to the specified key in memory, and persists this change in storage.
+     */
+    public function set(string $key, $value): void
+    {
+        $this->flashData[$key] = $value;
+        $this->persist();
+    }
+
+    /**
+     * @inheritDoc
+     *
+     * Removes the specified key from memory, and persists this change in storage.
+     */
+    public function unset(string $key): void
+    {
+        if (!array_key_exists($key, $this->flashData)) {
+            throw new NotFoundException(sprintf('Flash data not found for key "%1$s"', $key));
+        }
+
+        unset($this->flashData[$key]);
+        $this->persist();
+    }
+
+    /**
+     * @inheritDoc
+     *
+     * Clears all of this instance's data from memory.
+     */
+    public function clear(): void
```

Comment:
Making it clearable doesn't sound very useful. Can be confusing that if you call it after `set` it will be cleared only from memory.

Reply:
As with many other containers, including session, it may be necessary to remove all data from that container. Container data members, however, are not necessarily enumerable, such as is the case here. In such situations, `clear()` is the only way. That's the use of the `clear()` method. I'm not sure why that's not useful.

As for calling it after `set()`, that's fine. Remember that consumers of this container only know that it's a mutable, clearable container. They don't know whether it stores data in memory, in DB, or on disk. What's important is consistency of the data, e.g. what you `set()` is what you `get()` etc. In the same way, what's important is that after you `clear()`, `has()`, `get()`, and `unset()` don't work until you `set()` something. This container is consistent with that. What else does the consumer need to know?

Reply:
Or are you trying to say that maybe `clear()` should clear from storage as well? 🤔

Reply:
If that's the case, then perhaps you are right. Initially, I thought that it's not necessary because data is cleared from storage on `init()`, but the scenario with calling `clear()` after `set()` is not something I had considered. Would you agree that it should clear data from storage?

Reply:
After all, it's a mutator method, and all other mutator methods consistently persist the data in storage.

Reply:
Not sure when `clear` may be needed for flash data, but if it exists it probably should persist changes too, to make it consistent.

Reply:
Consumers of this class don't know that it represents flash data. They know that it is a mutable container. As such, it is useful to clear it.

But you're right. I'll make it persistent.

Reply:
AlexP11223, done. Does everything look OK now?

Reply:
Yeah. :)